packages/http-signature-utils/src/utils/verification.ts

```diff
@@ -0,0 +1,120 @@
+  if (
+    !sigInputComponents ||
+    !validateSigInputComponents(sigInputComponents, request)
+  )
+    return false
+
+  return (
+    sig && sigInput && typeof sig === 'string' && typeof sigInput === 'string'
+  )
+}
+
+export async function verifySigAndChallenge(
+  clientKey: JWK,
+  request: RequestLike
+): Promise<boolean> {
+  const sig = request.headers['signature'] as string
+  const sigInput = request.headers['signature-input'] as string
+  const challenge = sigInputToChallenge(sigInput, request)
+  if (!challenge) {
+    return false
+  }
+
+  return await verifySig(sig.replace('sig1=', ''), clientKey, challenge)
+}
+
+//exported for tests
+export async function verifySig(
+  sig: string,
+  jwk: JWK,
+  challenge: string
+): Promise<boolean> {
+  const publicKey = (await importJWK(jwk)) as crypto.KeyLike
+  const data = Buffer.from(challenge)
+  return crypto.verify(null, data, publicKey, Buffer.from(sig, 'base64'))
+}
+
+//exported for tests
+export function sigInputToChallenge(
+  sigInput: string,
+  request: RequestLike
+): string | null {
+  const sigInputComponents = getSigInputComponents(sigInput)
+
+  if (
+    !sigInputComponents ||
+    !validateSigInputComponents(sigInputComponents, request)
+  )
+    return null
+
+  // https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-09#section-2.3
+  let signatureBase = ''
+  for (const component of sigInputComponents) {
+    if (component === '@method') {
+      signatureBase += `"@method": ${request.method}\n`
+    } else if (component === '@target-uri') {
+      signatureBase += `"@target-uri": ${request.url}\n`
+    } else {
+      signatureBase += `"${component}": ${request.headers[component]}\n`
+    }
+  }
+
+  signatureBase += `"@signature-params": ${(
+    request.headers['signature-input'] as string
+  )?.replace('sig1=', '')}`
+  return signatureBase
+}
+
+// TODO: Replace with public httpsig library
+function getSigInputComponents(sigInput: string): string[] | null {
+  // https://datatracker.ietf.org/doc/html/rfc8941#section-4.1.1.1
+  const messageComponents = sigInput
+    .split('sig1=')[1]
+    ?.split(';')[0]
+    ?.split(' ')
+  return messageComponents
+    ? messageComponents.map((component) => component.replace(/[()"]/g, ''))
+    : null
+}
+
+function validateSigInputComponents(
+  sigInputComponents: string[],
+  request: RequestLike
+): boolean {
+  // https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol#section-7.3.1
+
+  for (const component of sigInputComponents) {
+    // https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-09#section-2.1
+    if (component !== component.toLowerCase()) return false
+  }
+
+  const isValidContentDigest =
+    !sigInputComponents.includes('content-digest') ||
+    (!!request.headers['content-digest'] &&
+      request.body &&
+      Object.keys(request.body).length > 0 &&
+      sigInputComponents.includes('content-digest') &&
```

Comment:
Should we require that `content-length` and `content-type` are also included?

Reply:
In the headers, in the `Signature-Input` components, in both?

Reply:
Both unless they're somehow not required by GNAP/httbis?
I think `content-type` is being enforced in the headers by the OpenAPI spec's:
```
      requestBody:
        content:
          application/json:
```